pkg/action/lazyclient.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package action
+
+import (
+	"context"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type lazyClient struct {
```

Comment:
why is it called lazyClient? because it only runs init() once used?

Reply:
It only initializes the underlying client when it must contact the API server.

Kubernetes 1.18 introduced a breaking change in the way they load the kubeconfig. It now contacts the API server before any user-requested operations are executed.

Helm has traditionally loaded the Kubeconfig very early, and passed the client around from outside via the normal dependency injection model. With this change, we had to figure out a way to not change or break the way the Helm client is created, but defer the underlying Kubernetes initialization until as late as possible. We did this by "lazy loading" the client in the `lazyClient`.

This was all done in the name of preserving the same API for Helm API users (as opposed to the other suggestion earlier in the thread, which would have required downstream users to rewrite their code as well).

Reply:
> Kubernetes 1.18 introduced a breaking change in the way they load the kubeconfig. It now contacts the API server before any user-requested operations are executed.

It actually doesn't contact the server, but it does stop silently defaulting to an insecure configuration when asked to load a non-existent configuration.

Reply:
You are right. Net result: Same thing.

Reply:
Changing behavior is a breaking change.  Unlike Kubernetes, Helm follows semantic versioning.  We need to figure out work arounds when Kubernetes breaks backwards compatibility.